sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/serializer/TypeReference.java

```diff
@@ -24,19 +29,44 @@
      * @throws IllegalArgumentException If the reference is constructed without type information.
      */
     public TypeReference() {
-        java.lang.reflect.Type superClass = this.getClass().getGenericSuperclass();
+        Type superClass = this.getClass().getGenericSuperclass();
         if (superClass instanceof Class) {
             throw LOGGER.logExceptionAsError(new IllegalArgumentException(MISSING_TYPE));
         } else {
             this.javaType = ((ParameterizedType) superClass).getActualTypeArguments()[0];
         }
     }
 
+    private TypeReference(Class<T> clazz) {
+        this.javaType = clazz;
+    }
+
     /**
-     * Return class T type
-     * @return type
+     * Returns the {@link Type} representing {@code T}.
+     *
+     * @return The {@link Type} representing {@code T}.
      */
-    public java.lang.reflect.Type getJavaType() {
+    public Type getJavaType() {
         return javaType;
     }
+
+    /**
+     * Creates and instance of {@link TypeReference} which maintains the generic {@code T} of the passed {@link Class}.
+     * <p>
+     * This method will cache the instance of {@link TypeReference} using the passed {@link Class} as the key. This is
+     * meant to be used with non-generic types such as primitive object types and POJOs, not {@code Map<String, Object>}
+     * or {@code List<Integer>} parameterized types.
+     *
+     * @param clazz {@link Class} that contains generic information used to create the {@link TypeReference}.
+     * @param <T> The generic type.
+     * @return Either the cached or new instance of {@link TypeReference}.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> TypeReference<T> createInstance(Class<T> clazz) {
```

Comment:
It looks a little weird by having both createInstance and public constructor.

Reply:
They serve different purposes. The public constructor is for anonymous instances that use parameterized types for their `T`, such as `Map<String, String>`. This factory method is meant for non-parameterized `Class` types that can be safely cached and maintain `T`.

Reply:
Yeah, this is a bit odd. They serve different purposes but the constructor works for all cases and if the caller always uses the ctor not knowing about the static `createInstance` method, it may defeat the purpose of caching.
Also, in azure-core, there is a TypeUtil that has a utility method to `createParameterizedType()` that may be useful.

Reply:
The public constructor work with most of the cases like (concrete class, paramterized type), but not the clazz variable cases.
E.g.
```
public void someMethod(Class<T> clazz) {
    TypeReference<T> type = new TypeReference<T>() {} // type is null in this case, no matter what clazz you passed in.
    ...
 }
```

In this case, we must use the `createInstance` API. However, createInstance API does not serve all purpose as well, we cannot pass the parameterized type here.

So both APIs have hard limitations and caching strategy are different behind the API. If I am not expert users I would confuse of the usage here.

Reply:
Agreed @sima-zhu, having both does lead to some confusion. Luckily, this type is mainly going to be used internally where determining whether to use `createInstance(Class<T>)` or the constructor will be based on the API. If the type is known ahead of time and it is parameterized the only choice is to use the constructor, if the type is known but is a `Class<T>` or is passed into the API `createInstance(Class<T>)` will be used.